src/Npgsql/NpgsqlDataSource.cs

```diff
@@ -503,6 +503,7 @@ protected virtual async ValueTask DisposeAsyncBase()
             await _periodicPasswordProviderTimer.DisposeAsync().ConfigureAwait(false);
 
         _setupMappingsSemaphore.Dispose();
+        MetricsReporter.Dispose(); // TODO: This is probably too early, dispose only when all connections have been closed?
```

Comment:
I think it's OK... This means you would stop getting metrics as soon as the data source instance is disposed (as opposed to when the last connection has been closed), right? If so, sounds like very reasonable behavior to me (I'd even remove the TODO...).